### PR TITLE
Yosys Step Rewrite, Update `nix-eda`

### LINKED ISSUE #221: Investigate Python API for Yosys
Yosys now has a Python API. It may be prudent to migrate Yosys-based scripts to Python from Tcl for codebase consistency.

### PR DESCRIPTION
**Pull request by donn**
* `Yosys.JsonHeader`, `Yosys.*Synthesis`
  * **Internal**: * Steps are no longer `TclStep`s: rewritten in Python and now use `libyosys`. While there are no functional changes, this enhances the codebase's consistency and helps avoid tokenization-related security issues.

* `Yosys.*Synthesis`
  * ABC scripts used now created dynamically and dumped as a `.abc` file into the step directory.
  * Implemented many of the suggestions in https://github.com/efabless/openlane2/issues/524, some behind flags:
    * `SYNTH_ABC_DFF`: Adds `-dff` to `abc` invocations (except the ones inside `synth`)
    * `SYNTH_ABC_BOOTH`: Activates the [`booth`](https://yosyshq.readthedocs.io/projects/yosys/en/0.44/cmd/booth.html) pass as part of `synth`
    * `SYNTH_ABC_USE_MFS3`: Uses `mfs3` in all strategies before retime
    * `SYNTH_ABC_AREA_USE_NF`: Attempts delay-based mapping with a really high delay value instead of area-based mapping.

## Tool Updates

* Updated `nix-eda`
  * `yosys` -> `0.44` (+ `-y` patch)
  * `klayout` -> `0.29.4`
  * OpenROAD now used with new `withPythonPackages` features to use Python packages specifically for the OpenROAD environment
 
---
Supersedes #464 
Resolves #221

### CODE REVIEW

File: openlane/steps/step.py [hunk 1]
```diff
@@ -1266,7 +1266,7 @@ def run_subprocess(
             output_processors.append(cls(self, report_dir, silent))
 
         verbose(
-            f"Logging subprocess to [repr.filename][link=file://{os.path.abspath(log_path)}]{os.path.relpath(log_path)}[/link][/repr.filename]…"
+            f"Logging subprocess to [repr.filename]'{os.path.relpath(log_path)}'[/repr.filename]…"
```
Comment (kareefardi) [inline openlane/steps/step.py:1269]: why?

File: openlane/flows/synth_explore.py [hunk 2]
```diff
@@ -79,53 +79,62 @@ def run(
...
+                "AREA 2",
+                "AREA 3",
+                "DELAY 0",
+                "DELAY 1",
+                "DELAY 2",
+                "DELAY 3",
+                "DELAY 4",
+            ]:
+                config = self.config.copy(
+                    SYNTH_ABC_AREA_USE_NF=nf,
```
Comment (kareefardi) [inline openlane/flows/synth_explore.py:95]: not in changelog. also why?

Comment (donn) [reply openlane/flows/synth_explore.py:95]: mistake, will revert

**Comment by kareefardi** [comment]
Can we include https://github.com/The-OpenROAD-Project/OpenLane/pull/2100 as part of this PR?

**Review by kareefardi — APPROVED** [review]

**Pull request merged**